File: android/app/src/main/java/cat/plou/radar/Frames.kt
package cat.plou.radar

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/** Un fotograma de radar: instante y ruta base de sus teselas. */
data class RadarFrame(
    /** Instante del fotograma (epoch en segundos). */
    val time: Long,
    /** Ruta base de las teselas de ese fotograma. */
    val path: String,
    /** `past` = observación; `nowcast` = extrapolación del proveedor. */
    val nowcast: Boolean,
)

data class RadarIndex(
    /** Base de URL de las teselas. */
    val host: String,
    val generated: Long,
    val past: List<RadarFrame>,
    val forecast: List<RadarFrame>,
) {
    /** Todos los fotogramas en orden cronológico. */
    val all: List<RadarFrame> get() = past + forecast

    /** Fotograma observado más reciente. */
    val latest: RadarFrame? get() = past.lastOrNull()
}

/** Opciones de una tesela de radar. */
data class TileOptions(
    val size: Int = 512,
    val color: Int = ANALYSIS_SCHEME.id,
    val smooth: Boolean = false,
    val snow: Boolean = true,
)

private const val INDEX_URL = "https://api.rainviewer.com/public/weather-maps.json"
private const val USER_AGENT = "Plou/1.0 (+https://github.com/borborborja/plou)"

/**
 * Cliente del índice de fotogramas. La app habla directamente con el proveedor:
 * no hay servidor propio de por medio.
 */
class RadarIndexClient(
    private val http: OkHttpClient = defaultHttpClient(),
    private val ttlMillis: Long = 120_000,
    private val now: () -> Long = System::currentTimeMillis,
) {
    private var cached: RadarIndex? = null
    private var fetchedAt = 0L

    private val json = Json { ignoreUnknownKeys = true }

    /** Índice cacheado durante [ttlMillis]; ante un fallo sirve el último bueno. */
    @Synchronized
    fun get(force: Boolean = false): RadarIndex {
        val hit = cached
        if (!force && hit != null && now() - fetchedAt < ttlMillis) return hit
        return try {
            val fresh = load()
            cached = fresh
            fetchedAt = now()
            fresh
        } catch (e: Exception) {
            hit ?: throw e
        }
    }

    private fun load(): RadarIndex {
        val request = Request.Builder()
            .url(INDEX_URL)
            .header("User-Agent", USER_AGENT)
            .header("Accept", "application/json")
            .build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Índice de radar: HTTP ${response.code}")
            val body = response.body?.string() ?: error("Índice de radar vacío")
            return parseIndex(body)
        }
    }

    internal fun parseIndex(body: String): RadarIndex {
        val root = json.parseToJsonElement(body).jsonObject
        val host = root["host"]?.jsonPrimitive?.content?.trimEnd('/')
            ?: error("Índice de radar sin campo `host`")
        val radar = root["radar"]?.jsonObject
        val past = frames(radar, "past", nowcast = false)
        val forecast = frames(radar, "nowcast", nowcast = true)
        if (past.isEmpty()) error("Índice de radar sin fotogramas")
        return RadarIndex(
            host = host,
            generated = root["generated"]?.jsonPrimitive?.content?.toLongOrNull()
                ?: (System.currentTimeMillis() / 1000),
            past = past,
            forecast = forecast,
        )
    }

    private fun frames(
        parent: kotlinx.serialization.json.JsonObject?,
        key: String,
        nowcast: Boolean,
    ): List<RadarFrame> {
        val array = parent?.get(key)?.jsonArray ?: return emptyList()
        return array.mapNotNull { element ->
            val obj = element.jsonObject
            val time = obj["time"]?.jsonPrimitive?.content?.toLongOrNull() ?: return@mapNotNull null
            val path = obj["path"]?.jsonPrimitive?.content ?: return@mapNotNull null
            RadarFrame(time, path, nowcast)
        }.sortedBy { it.time }
    }
}

/**
 * URL de una tesela.
 * Formato: `{host}{path}/{size}/{z}/{x}/{y}/{color}/{smooth}_{snow}.png`
 */
fun tileUrl(index: RadarIndex, frame: RadarFrame, z: Int, x: Int, y: Int, opts: TileOptions): String {
    val smooth = if (opts.smooth) 1 else 0
    val snow = if (opts.snow) 1 else 0
    return "${index.host}${frame.path}/${opts.size}/$z/$x/$y/${opts.color}/${smooth}_$snow.png"
}

/** Plantilla con los marcadores que espera un mapa de teselas. */
fun tileUrlTemplate(index: RadarIndex, frame: RadarFrame, opts: TileOptions): String {
    val smooth = if (opts.smooth) 1 else 0
    val snow = if (opts.snow) 1 else 0
    return "${index.host}${frame.path}/${opts.size}/%d/%d/%d/${opts.color}/${smooth}_$snow.png"
}

fun defaultHttpClient(): OkHttpClient = OkHttpClient.Builder()
    .connectTimeout(10, TimeUnit.SECONDS)
    .readTimeout(15, TimeUnit.SECONDS)
    .build()

File: server/src/radar/frames.ts
import { config } from '../config.js';

export interface RadarFrame {
  /** Instante del fotograma (epoch en segundos). */
  time: number;
  /** Ruta base de las teselas de ese fotograma. */
  path: string;
  /** `past` = observación; `nowcast` = extrapolación del proveedor. */
  kind: 'past' | 'nowcast';
}

export interface RadarIndex {
  /** Base de URL de las teselas. */
  host: string;
  generated: number;
  past: RadarFrame[];
  nowcast: RadarFrame[];
  infrared: RadarFrame[];
}

interface RawFrame {
  time?: number;
  path?: string;
}

interface RawIndex {
  host?: string;
  generated?: number;
  radar?: { past?: RawFrame[]; nowcast?: RawFrame[] };
  satellite?: { infrared?: RawFrame[] };
}

function mapFrames(raw: RawFrame[] | undefined, kind: RadarFrame['kind']): RadarFrame[] {
  if (!Array.isArray(raw)) return [];
  return raw
    .filter((f): f is Required<RawFrame> => typeof f.time === 'number' && typeof f.path === 'string')
    .map((f) => ({ time: f.time, path: f.path, kind }))
    .sort((a, b) => a.time - b.time);
}

let cached: { index: RadarIndex; fetchedAt: number } | null = null;
let inFlight: Promise<RadarIndex> | null = null;

async function load(): Promise<RadarIndex> {
  const res = await fetch(config.radar.indexUrl, {
    headers: { 'user-agent': config.userAgent, accept: 'application/json' },
    signal: AbortSignal.timeout(config.radar.tileTimeoutMs),
  });
  if (!res.ok) throw new Error(`Índice de radar: HTTP ${res.status}`);
  const raw = (await res.json()) as RawIndex;
  if (!raw.host) throw new Error('Índice de radar sin campo `host`');

  const index: RadarIndex = {
    host: raw.host.replace(/\/$/, ''),
    generated: raw.generated ?? Math.floor(Date.now() / 1000),
    past: mapFrames(raw.radar?.past, 'past'),
    nowcast: mapFrames(raw.radar?.nowcast, 'nowcast'),
    infrared: mapFrames(raw.satellite?.infrared, 'past'),
  };
  if (index.past.length === 0) throw new Error('Índice de radar sin fotogramas');
  cached = { index, fetchedAt: Date.now() };
  return index;
}

/** Índice de fotogramas, cacheado durante `radar.indexTtlSeconds`. */
export async function getRadarIndex(force = false): Promise<RadarIndex> {
  const ttl = config.radar.indexTtlSeconds * 1000;
  if (!force && cached && Date.now() - cached.fetchedAt < ttl) return cached.index;
  if (inFlight) return inFlight;
  inFlight = load().finally(() => {
    inFlight = null;
  });
  try {
    return await inFlight;
  } catch (err) {
    // Ante un fallo puntual se sirve el último índice conocido si aún es usable.
    if (cached) return cached.index;
    throw err;
  }
}

/** Todos los fotogramas en orden cronológico (observación + extrapolación). */
export function allFrames(index: RadarIndex): RadarFrame[] {
  return [...index.past, ...index.nowcast];
}

/** Fotograma observado más reciente. */
export function latestPast(index: RadarIndex): RadarFrame | undefined {
  return index.past.at(-1);
}

/**
 * Construye la URL de una tesela.
 *
 * Formato: `{host}{path}/{size}/{z}/{x}/{y}/{color}/{smooth}_{snow}.png`
 */
export function tileUrl(
  index: RadarIndex,
  frame: RadarFrame,
  opts: { z: number; x: number; y: number; size?: number; color?: number; smooth?: boolean; snow?: boolean },
): string {
  const size = opts.size ?? config.radar.tileSize;
  const color = opts.color ?? 2;
  const smooth = opts.smooth ? 1 : 0;
  const snow = opts.snow ? 1 : 0;
  return `${index.host}${frame.path}/${size}/${opts.z}/${opts.x}/${opts.y}/${color}/${smooth}_${snow}.png`;
}

/**
 * Plantilla de URL con los marcadores `{z}/{x}/{y}` que espera un mapa de
 * teselas, para que el cliente no tenga que conocer el formato del proveedor.
 */
export function tileUrlTemplate(
  index: RadarIndex,
  frame: RadarFrame,
  opts: { size?: number; color?: number; smooth?: boolean; snow?: boolean } = {},
): string {
  const size = opts.size ?? config.radar.tileSize;
  const color = opts.color ?? 2;
  const smooth = opts.smooth ? 1 : 0;
  const snow = opts.snow ? 1 : 0;
  return `${index.host}${frame.path}/${size}/{z}/{x}/{y}/${color}/${smooth}_${snow}.png`;
}

/** Plantilla equivalente para la capa de cobertura. */
export function coverageTileUrlTemplate(index: RadarIndex, size?: number): string {
  const s = size ?? config.radar.tileSize;
  return `${index.host}/v2/coverage/0/${s}/{z}/{x}/{y}/0/0_0.png`;
}

/** URL de la capa de cobertura de radar (dónde hay datos disponibles). */
export function coverageTileUrl(
  index: RadarIndex,
  opts: { z: number; x: number; y: number; size?: number },
): string {
  const size = opts.size ?? config.radar.tileSize;
  return `${index.host}/v2/coverage/0/${size}/${opts.z}/${opts.x}/${opts.y}/0/0_0.png`;
}

/** Utilidad de test: vacía la caché del índice. */
export function resetRadarIndexCache(): void {
  cached = null;
  inFlight = null;
}
